### PR TITLE
Pie chart feature

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: addon/templates/components/ember-sparkles/bar-chart.hbs [hunk 1]
```diff
@@ -47,20 +47,20 @@
...
-        (d3-style 'fill' 'steelblue')
-        (d3-style 'opacity' 0)
-        (if with-transition (d3-transition transition) (d3-noop))
-        (d3-style 'opacity' 1)
-        (d3-attr 'x' (pipe inputAccessor xScale))
-        (d3-attr 'width' xScale.bandwidth)
-        (d3-attr 'y' (pipe outputAccessor yScale))
-        (d3-attr 'height' (pipe outputAccessor yScale (bar-height height)))
-      )
+    enter=(pipe
```
Comment (ghost) [inline addon/templates/components/ember-sparkles/bar-chart.hbs]: can you please revert the changes to this file? they shouldn't be in this PR

File: addon/components/ember-sparkles/pie-chart.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import Ember from 'ember';
+import layout from '../../templates/components/ember-sparkles/pie-chart';
+
+export default Ember.Component.extend({
+  layout,
+  radius: Math.min(this.width, this.height) / 2,
```
Comment (taras) [inline addon/components/ember-sparkles/pie-chart.js]: Should this be a computed property?

File: tests/dummy/app/templates/pie-chart.hbs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{{!-- {{#ember-sparkles as |chart|}}
```
Comment (taras) [inline tests/dummy/app/templates/pie-chart.hbs]: Why is this commented out?

File: addon/components/ember-sparkles/pie-chart.js (hunk 2 above)
Comment (ghost) [inline addon/components/ember-sparkles/pie-chart.js]: this will most likely not live in any JS file once we get there, Albert is working on this branch to build a pie chart, he has just begun.

Comment (taras) [inline addon/components/ember-sparkles/pie-chart.js]: Ok

File: addon/templates/components/ember-sparkles/bar-chart.hbs (hunk 1 above)
Comment (guozhaonan) [inline addon/templates/components/ember-sparkles/bar-chart.hbs]: yep, will resolve in the next commit

File: tests/dummy/app/templates/pie-chart.hbs (hunk 3 above)
Comment (guozhaonan) [inline tests/dummy/app/templates/pie-chart.hbs]: I commented out the tests section because i was a bit overzealous when I started and generated files and started writing code. I commented out the code so that the build doesn't break while I write the add-on, right now I'm going to finish writing the add-on and then start working on the test.

**Comment by ghost** [comment]
addresses #13

File: addon/helpers/ember-sparkles/arc.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import Ember from 'ember';
+import { arc } from 'd3-shape';
+
+export function emberSparklesArc([ radius ], textLabel=false) {
+  if (textLabel === true){
+    return arc().outerRadius(radius - 40).innerRadius(radius - 40);
```
Comment (ghost) [inline addon/helpers/ember-sparkles/arc.js]: we will want to pass `outerRadius` and `innerRadius` arguments to this helper instead of hardcoding their inputs

Comment (ghost) [inline addon/helpers/ember-sparkles/arc.js]: this means we will be able to remove the `if (textLabel...)` part (since the arguments will be different depending on whether these are text labels or not)

File: addon/helpers/ember-sparkles/pie.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import Ember from 'ember';
+import { pie } from 'd3-shape';
+
+export function emberSparklesPie(params, { valueFn }) {
+  return pie().sortValues(function(a, b) { return a - b; }).value(valueFn);
```
Comment (ghost) [inline addon/helpers/ember-sparkles/pie.js]: a bit more concisely:
`return pie().sortValues((a, b) => a - b).value(valueFn);`

File: tests/unit/helpers/ember-sparkles/pie-test.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      key: 'arc 3',
+      value: 50
+    }
+  ];
+
+  let valueFn = ({ value }) => value;
+
+  let pie = emberSparklesPie([], { valueFn });
+  let arrayOfArcs = pie(data);
+  let arrayOfAngles = pie(data).map(object => {
```
Comment (ghost) [inline tests/unit/helpers/ember-sparkles/pie-test.js]: you can clean this up a bit by doing this:

```
let pieData = pie(data);

let arcs = pieData.map(({ startAngle, endAngle }) => (endAngle - startAngle).toFixed(6));
```

File: tests/unit/helpers/ember-sparkles/pie-test.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    return (object.endAngle - object.startAngle).toFixed(6);
+  });
+
+  assert.ok(arrayOfArcs);
```
Comment (ghost) [inline tests/unit/helpers/ember-sparkles/pie-test.js]: this test is redundant with the second test

Comment (guozhaonan) [inline tests/unit/helpers/ember-sparkles/pie-test.js]: the first test or the third test makes the second test redundant?

File: addon/helpers/ember-sparkles/arc-translate.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+export function emberSparklesArcTranslate([ centroid ], { height, width }) {
+  return function(d) {
+    let x, y;
+    [x, y] = centroid(d)
+    x = x + width;
+    y = y + height;
+    let coordinates = [];
+    coordinates.push(x);
+    coordinates.push(y);
+    return `translate(${coordinates})`;
```
Comment (ghost) [inline addon/helpers/ember-sparkles/arc-translate.js]: let's talk about this part - you're mutating some variables, and we don't need to use the array.

File: addon/templates/components/ember-sparkles/pie-chart.hbs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+              (d3-attr 'class' 'arcTextStyle')
+              (d3-attr 'dy' '.35em')
+              (d3-attr 'transform' (ember-sparkles/arc-translate constructors.arc.centroid height=centerHeight width=centerWidth))
+              (if with-transition (d3-transition transition) (d3-noop))
+              (d3-text (pipe (d3-get 'data.key')))
+              (d3-style 'text-anchor' 'black')
+            )
+            update=(pipe
+              (d3-style 'fill' 'black')
+              (d3-attr 'class' 'arcTextStyle')
```
Comment (ghost) [inline addon/templates/components/ember-sparkles/pie-chart.hbs]: you can remove the setting of the class here (any selections in the update method will already have that class)

Comment (guozhaonan) [inline addon/templates/components/ember-sparkles/pie-chart.hbs]: Oh, I mainly did that so that I could set the text styling
update: I see what you mean, fixing it in the next push

File: package.json [hunk 10]
```diff
@@ -56,7 +56,7 @@
     "ember-cli-babel": "^5.1.6",
     "ember-cli-d3-shape": "0.9.0",
     "ember-cli-htmlbars": "^1.0.3",
-    "ember-d3-helpers": "0.5.1",
+    "ember-d3-helpers": "git@github.com:LocusEnergy/ember-d3-helpers.git",
```
Comment (ghost) [inline package.json]: let's point this to v0.5.2 (which I just published)

File: tests/dummy/app/controllers/pie-chart.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import Ember from 'ember';
+import _ from 'lodash/lodash';
+
+let generateData = function() {
+  let keys = [];
+//   let keys = [
```
Comment (ghost) [inline tests/dummy/app/controllers/pie-chart.js]: please delete commented code

File: addon/helpers/ember-sparkles/get-percentage.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import Ember from 'ember';
```
Comment (ghost) [inline addon/helpers/ember-sparkles/get-percentage.js]: you can delete this file

Comment (guozhaonan) [inline addon/helpers/ember-sparkles/get-percentage.js]: done

**Comment by ghost** [comment]
@taras can you review this PR please? The pie chart is fully functional now

File: tests/integration/components/ember-sparkles/pie-chart-test.js [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    }, {
+      key: 'arc 4',
+      value: 50
+    }
+  ];
+  domain = data.map(({ key }) => key);
+
+  this.setProperties({ data, domain });
+
+  assert.equal(this.$('path').length, 4, 'There are the correct number of arcs after the data is updated');
```
Comment (ghost) [inline tests/integration/components/ember-sparkles/pie-chart-test.js:117]: I know that checking the data bound to each arc is very hard to do - instead, can you also check the arc percentages as a proxy for this?

File: addon/components/ember-sparkles/pie-chart.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+export default Ember.Component.extend({
+  layout,
+  tagName: 'g',
+  classNames: ['ember-sparkles--pie-chart'],
+  'with-transition': true,
+  'with-arc-labels': false,
+
+  didInsertElement() {
+    this._super(...arguments);
+    Ember.run.scheduleOnce('afterRender', this, 'renderChart');
```
Comment (taras) [inline addon/components/ember-sparkles/pie-chart.js:14]: We should probably make this a mixin.

File: addon/templates/components/ember-sparkles/pie-chart.hbs [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+    {{shhh (compute (d3-join 'text' domain
+        enter=(pipe
+          (d3-append 'text')
+          (d3-attr 'opacity' 0)
+          (d3-attr 'dx' '-3.8em')
+          (d3-attr 'dy' '.35em')
+          (d3-style 'font-size' '10px')
+          (if with-transition (d3-transition transition) (d3-noop))
+          (d3-text (d3-get))
```
Comment (taras) [inline addon/templates/components/ember-sparkles/pie-chart.hbs]: Does this do anything?

**Comment by taras** [comment]
I get this error when I click around the pie chart. 

screen shot 2016-07-18 at 3 19 33 am

I did some googling. It looks like you need a custom tween for this. http://stackoverflow.com/a/21287079/172894

File: addon/templates/components/ember-sparkles/pie-chart.hbs (hunk 15 above)
Comment (taras) [inline addon/templates/components/ember-sparkles/pie-chart.hbs]: I think I get what it's doing. `d3-get` should be 2 separate helpers. One that return the 1st param and the other that gets the value. I added `r/get` helper to `ember-reactive-helpers` to handle the 2nd use case. I'm going to make another helper called `r/param` which you can use to take a take a param. We should use that here instead.

Comment (taras) [inline addon/templates/components/ember-sparkles/pie-chart.hbs]: I've added both of these to `ember-reactive-helpers` we should refactor the app to use these new reactive helpers and remove `d3-get` from `ember-d3-helpers`

File: addon/components/ember-sparkles/pie-chart.js (hunk 14 above)
Comment (ghost) [inline addon/components/ember-sparkles/pie-chart.js:14]: a mixin or a base component class that we extend (the latter is what I'm doing on the axes refactor branch). It seems like base component pattern is a bit more Ember-esque; what do you think?

File: addon/templates/components/ember-sparkles/pie-chart.hbs (hunk 15 above)
Comment (ghost) [inline addon/templates/components/ember-sparkles/pie-chart.hbs]: Oh, this is great! Thanks Taras

File: addon/components/ember-sparkles/pie-chart.js (hunk 14 above)
Comment (ghost) [inline addon/components/ember-sparkles/pie-chart.js:14]: Like this:
https://github.com/LocusEnergy/ember-sparkles/blob/area-helper/addon/components/ember-sparkles/d3-element.js

used like: https://github.com/LocusEnergy/ember-sparkles/blob/area-helper/addon/components/ember-sparkles/bar-chart.js

**Comment by ghost** [comment]
Looks great!

**Comment by ghost** [comment]
closes #3
